checkpoint/side_handler.go

```diff
@@ -124,6 +167,78 @@ func NewPostTxHandler(k Keeper, contractCaller helper.IContractCaller) hmTypes.P
 	}
 }
 
+// PostHandleMsgCheckpointAdjust handles msg checkpoint adjust
+func PostHandleMsgCheckpointAdjust(ctx sdk.Context, k Keeper, msg types.MsgCheckpointAdjust, sideTxResult abci.SideTxResultType, contractCaller helper.IContractCaller) sdk.Result {
+	logger := k.Logger(ctx)
+	chainParams := k.ck.GetParams(ctx).ChainParams
+	params := k.GetParams(ctx)
+
+	// Skip handler if checkpoint-adjust is not approved
+	if sideTxResult != abci.SideTxResultType_Yes {
+		logger.Debug("Skipping new checkpoint-adjust since side-tx didn't get yes votes", "checkpointNumber", msg.HeaderIndex)
+		return common.ErrBadBlockDetails(k.Codespace()).Result()
+	}
+
+	checkpointBuffer, err := k.GetCheckpointFromBuffer(ctx)
+	if checkpointBuffer != nil {
+		logger.Error("checkpoint buffer", "error", err)
+		return common.ErrNoCheckpointFound(k.Codespace()).Result()
+	}
+
+	checkpointObj, err := k.GetCheckpointByNumber(ctx, msg.HeaderIndex)
+	if err != nil {
+		logger.Error("Unable to get checkpoint from db", "error", err)
+		return common.ErrNoCheckpointFound(k.Codespace()).Result()
+	}
+
+	rootChainInstance, err := contractCaller.GetRootChainInstance(chainParams.RootChainAddress.EthAddress())
+	if err != nil {
+		logger.Error("Unable to fetch rootchain contract instance", "error", err)
+		return common.ErrNoCheckpointFound(k.Codespace()).Result()
+	}
+
+	root, start, end, _, _, err := contractCaller.GetHeaderInfo(msg.HeaderIndex, rootChainInstance, params.ChildBlockInterval)
+	if err != nil {
+		logger.Error("Unable to fetch checkpoint from rootchain", "error", err, "checkpointNumber", msg.HeaderIndex)
+		return common.ErrNoCheckpointFound(k.Codespace()).Result()
+	}
```

Comment:
This call shouldn't be here. We cannot call external sources from PostHandler, only in SideHandler. So we will check if msg is the same as data on Ethereum chain in Side handler and in PostHandler we will validate data same as Handler and store checkpointObj, that's it. 

```suggestion
```